// convert-svg-react/lib/components/Button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var Button = function Button() {
  var copySVGstring = function copySVGstring() {
    var svgText = document.getElementsByTagName('textarea')[1];
    svgText.select();
    document.execCommand('copy');
  };

  return /*#__PURE__*/_react["default"].createElement('button', {
    key: 'button',
    onClick: function onClick() {
      copySVGstring();
    },
    style: {
      position: 'absolute',
      left: '150px',
      bottom: '40px',
      height: '40px',
      border: 'none',
      padding: '0 10px',
      borderRadius: '2px',
      cursor: 'pointer',
      background: '#dac0a7',
      color: '#fff',
      fontSize: '15px',
      fontWeight: '600',
      letterSpacing: '.025em',
      boxShadow: '0 4px 6px rgba(50,50,93,.11), 0 1px 3px rgba(0,0,0,.08)'
    }
  }, 'COPY IN CLIPBOARD');
};

var _default = Button;
exports["default"] = _default;
